"mismatched types" on indexing into trait with two Index subtraits via output type

If a trait has two or more `Index` subtraits, and one obtains a reference to something that implements that trait under specific conditions, typeck reports an error for indexing with all but the last `Index`'s key. Seems to happen only if that reference was obtained through another trait's output type.

Consider the following example ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=619415d35b0211d5daffd055f90593ae)):
```rust
use std::ops::Index;

struct KeyA;
struct KeyB;
struct KeyC;

trait Foo: Index<KeyA> + Index<KeyB> + Index<KeyC> {}
trait FooBuilder {
    type Inner: Foo;
    fn inner(&self) -> &Self::Inner;
}

fn do_stuff(foo: &impl FooBuilder) {
    let inner = foo.inner();
    &inner[KeyA];
    //     ^^^^ expected struct `KeyC`, found struct `KeyA`
    &inner[KeyB];
    //     ^^^^ expected struct `KeyC`, found struct `KeyB`
    &inner[KeyC];
}
```

This breaks with the message:
```
error[E0308]: mismatched types
  --> src/lib.rs:15:12
   |
15 |     &inner[KeyA];
   |            ^^^^ expected struct `KeyC`, found struct `KeyA`

error[E0308]: mismatched types
  --> src/lib.rs:16:12
   |
16 |     &inner[KeyB];
   |            ^^^^ expected struct `KeyC`, found struct `KeyB`
```

The issue does not occur if the indexing is performed on a `Foo` directly. That is, the following works just fine ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ede84fb6ba7fe32eda13633e99a14e68)):
```rust
fn do_stuff(foo: &impl Foo) {
    let inner = foo;
    &inner[KeyA];
    &inner[KeyB];
    &inner[KeyC];
}
```

What I would expect to happen is that all three index operations work. More specifically, I would expect each of the operations to produce the corresponding `Index::Output` type correctly.

### Meta

`rustc --version --verbose`:
```
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-unknown-linux-gnu
release: 1.42.0
LLVM version: 9.0
```

--- a/src/test/ui/generic-associated-types/issue-74816.rs
+++ b/src/test/ui/generic-associated-types/issue-74816.rs
@@ -1,0 +1,23 @@
+#![feature(associated_type_defaults)]
+#![feature(generic_associated_types)]
+#![allow(incomplete_features)]
+
+trait Trait1 {
+    fn foo();
+}
+
+trait Trait2 {
+    type Associated: Trait1 = Self;
+    //~^ ERROR: the trait bound `Self: Trait1` is not satisfied
+    //~| the size for values of type `Self` cannot be known
+}
+
+impl Trait2 for () {}
+
+fn call_foo<T: Trait2>() {
+    T::Associated::foo()
+}
+
+fn main() {
+    call_foo::<()>()
+}

--- a/src/test/ui/generic-associated-types/issue-74816.stderr
+++ b/src/test/ui/generic-associated-types/issue-74816.stderr
@@ -1,0 +1,31 @@
+error[E0277]: the trait bound `Self: Trait1` is not satisfied
+  --> $DIR/issue-74816.rs:10:5
+   |
+LL |     type Associated: Trait1 = Self;
+   |     ^^^^^^^^^^^^^^^^^------^^^^^^^^
+   |     |                |
+   |     |                required by this bound in `Trait2::Associated`
+   |     the trait `Trait1` is not implemented for `Self`
+   |
+help: consider further restricting `Self`
+   |
+LL | trait Trait2: Trait1 {
+   |             ^^^^^^^^
+
+error[E0277]: the size for values of type `Self` cannot be known at compilation time
+  --> $DIR/issue-74816.rs:10:5
+   |
+LL |     type Associated: Trait1 = Self;
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+   |     |
+   |     doesn't have a size known at compile-time
+   |     required by this bound in `Trait2::Associated`
+   |
+help: consider further restricting `Self`
+   |
+LL | trait Trait2: Sized {
+   |             ^^^^^^^
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0277`.

--- a/src/test/ui/impl-trait/issues/issue-70877.rs
+++ b/src/test/ui/impl-trait/issues/issue-70877.rs
@@ -1,0 +1,38 @@
+#![feature(type_alias_impl_trait)]
+#![feature(impl_trait_in_bindings)]
+#![allow(incomplete_features)]
+
+type FooArg<'a> = &'a dyn ToString;
+type FooRet = impl std::fmt::Debug;
+
+type FooItem = Box<dyn Fn(FooArg) -> FooRet>;
+type Foo = impl Iterator<Item = FooItem>; //~ ERROR: type mismatch
+
+#[repr(C)]
+struct Bar(u8);
+
+impl Iterator for Bar {
+    type Item = FooItem;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        Some(Box::new(quux))
+    }
+}
+
+fn quux(st: FooArg) -> FooRet {
+    Some(st.to_string())
+}
+
+fn ham() -> Foo {
+    Bar(1)
+}
+
+fn oof() -> impl std::fmt::Debug {
+    let mut bar = ham();
+    let func = bar.next().unwrap();
+    return func(&"oof");
+}
+
+fn main() {
+    let _ = oof();
+}

--- a/src/test/ui/impl-trait/issues/issue-70877.stderr
+++ b/src/test/ui/impl-trait/issues/issue-70877.stderr
@@ -1,0 +1,15 @@
+error[E0271]: type mismatch resolving `<Bar as Iterator>::Item == Box<(dyn for<'r> Fn(&'r (dyn ToString + 'r)) -> Option<String> + 'static)>`
+  --> $DIR/issue-70877.rs:9:12
+   |
+LL | type FooRet = impl std::fmt::Debug;
+   |               -------------------- the expected opaque type
+...
+LL | type Foo = impl Iterator<Item = FooItem>;
+   |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected opaque type, found enum `Option`
+   |
+   = note: expected struct `Box<(dyn for<'r> Fn(&'r (dyn ToString + 'r)) -> impl Debug + 'static)>`
+              found struct `Box<(dyn for<'r> Fn(&'r (dyn ToString + 'r)) -> Option<String> + 'static)>`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0271`.

--- a/src/test/ui/traits/issue-70944.rs
+++ b/src/test/ui/traits/issue-70944.rs
@@ -1,0 +1,23 @@
+// check-pass
+// Regression test of #70944, should compile fine.
+
+use std::ops::Index;
+
+pub struct KeyA;
+pub struct KeyB;
+pub struct KeyC;
+
+pub trait Foo: Index<KeyA> + Index<KeyB> + Index<KeyC> {}
+pub trait FooBuilder {
+    type Inner: Foo;
+    fn inner(&self) -> &Self::Inner;
+}
+
+pub fn do_stuff(foo: &impl FooBuilder) {
+    let inner = foo.inner();
+    &inner[KeyA];
+    &inner[KeyB];
+    &inner[KeyC];
+}
+
+fn main() {}

--- a/src/test/ui/traits/trait-alias/issue-75983.rs
+++ b/src/test/ui/traits/trait-alias/issue-75983.rs
@@ -1,0 +1,17 @@
+// check-pass
+
+#![feature(trait_alias)]
+
+struct Bar;
+trait Foo {}
+impl Foo for Bar {}
+
+trait Baz = Foo where Bar: Foo;
+
+fn new() -> impl Baz {
+    Bar
+}
+
+fn main() {
+    let _ = new();
+}

--- a/src/test/ui/unsized/issue-71659.rs
+++ b/src/test/ui/unsized/issue-71659.rs
@@ -1,15 +1,15 @@
 #![feature(unsize)]
 
 use std::marker::Unsize;
-use std::rc::Rc;
-use std::sync::Arc;
 
 pub trait CastTo<T: ?Sized>: Unsize<T> {
     fn cast_to(&self) -> &T;
-    fn cast_mut_to(&mut self) -> &mut T;
-    fn into_cast_to(self: Box<Self>) -> Box<T>;
-    fn cast_rc_to(self: Rc<Self>) -> Rc<T>;
-    fn cast_arc_to(self: Arc<Self>) -> Arc<T>;
+}
+
+impl<T: ?Sized, U: ?Sized + Unsize<T>> CastTo<T> for U {
+    fn cast_to(&self) -> &T {
+        self
+    }
 }
 
 impl<T: ?Sized> Cast for T {}
@@ -20,75 +20,13 @@ pub trait Cast {
     {
         self
     }
-
-    fn cast_mut<T>(&mut self) -> &mut T
-    where
-        Self: CastTo<T>,
-    {
-        self.cast_mut_to()
-    }
-
-    fn into_cast<T>(self: Box<Self>) -> Box<T>
-    where
-        Self: CastTo<T>,
-    {
-        self.into_cast_to()
-    }
-
-    fn cast_rc<T>(self: Rc<Self>) -> Rc<T>
-    where
-        Self: CastTo<T>,
-    {
-        self.cast_rc_to()
-    }
-
-    fn cast_arc<T>(self: Arc<Self>) -> Arc<T>
-    where
-        Self: CastTo<T>,
-    {
-        self.cast_arc_to()
-    }
-}
-impl<T: ?Sized, U: ?Sized + Unsize<T>> CastTo<T> for U {
-    fn cast_to(&self) -> &T {
-        self
-    }
-
-    fn cast_mut_to(&mut self) -> &mut T {
-        self
-    }
-
-    fn into_cast_to(self: Box<Self>) -> Box<T> {
-        self
-    }
-
-    fn cast_rc_to(self: Rc<Self>) -> Rc<T> {
-        self
-    }
-
-    fn cast_arc_to(self: Arc<Self>) -> Arc<T> {
-        self
-    }
 }
 
-pub trait Foo {
-    fn foo(&self) {
-        println!("Foo({})", core::any::type_name::<Self>());
-    }
-}
-
-pub trait Bar: CastTo<dyn Foo> + CastTo<dyn core::fmt::Debug> + CastTo<[i32]> {
-    fn bar(&self) {
-        println!("Bar({})", core::any::type_name::<Self>());
-    }
-}
-
-impl Foo for [i32; 10] {}
-impl Bar for [i32; 10] {}
+pub trait Foo: CastTo<[i32]> {}
+impl Foo for [i32; 0] {}
 
 fn main() {
-    let x = [0; 10];
-    let x: Box<dyn Bar> = Box::new(x);
-    let x = (*x).cast::<[i32]>();
-    //~^ ERROR: the trait bound `dyn Bar: CastTo<[i32]>` is not satisfied
+    let x: &dyn Foo = &[];
+    let x = x.cast::<[i32]>();
+    //~^ ERROR: the trait bound `dyn Foo: CastTo<[i32]>` is not satisfied
 }

--- a/src/test/ui/unsized/issue-71659.rs
+++ b/src/test/ui/unsized/issue-71659.rs
@@ -1,0 +1,94 @@
+#![feature(unsize)]
+
+use std::marker::Unsize;
+use std::rc::Rc;
+use std::sync::Arc;
+
+pub trait CastTo<T: ?Sized>: Unsize<T> {
+    fn cast_to(&self) -> &T;
+    fn cast_mut_to(&mut self) -> &mut T;
+    fn into_cast_to(self: Box<Self>) -> Box<T>;
+    fn cast_rc_to(self: Rc<Self>) -> Rc<T>;
+    fn cast_arc_to(self: Arc<Self>) -> Arc<T>;
+}
+
+impl<T: ?Sized> Cast for T {}
+pub trait Cast {
+    fn cast<T: ?Sized>(&self) -> &T
+    where
+        Self: CastTo<T>,
+    {
+        self
+    }
+
+    fn cast_mut<T>(&mut self) -> &mut T
+    where
+        Self: CastTo<T>,
+    {
+        self.cast_mut_to()
+    }
+
+    fn into_cast<T>(self: Box<Self>) -> Box<T>
+    where
+        Self: CastTo<T>,
+    {
+        self.into_cast_to()
+    }
+
+    fn cast_rc<T>(self: Rc<Self>) -> Rc<T>
+    where
+        Self: CastTo<T>,
+    {
+        self.cast_rc_to()
+    }
+
+    fn cast_arc<T>(self: Arc<Self>) -> Arc<T>
+    where
+        Self: CastTo<T>,
+    {
+        self.cast_arc_to()
+    }
+}
+impl<T: ?Sized, U: ?Sized + Unsize<T>> CastTo<T> for U {
+    fn cast_to(&self) -> &T {
+        self
+    }
+
+    fn cast_mut_to(&mut self) -> &mut T {
+        self
+    }
+
+    fn into_cast_to(self: Box<Self>) -> Box<T> {
+        self
+    }
+
+    fn cast_rc_to(self: Rc<Self>) -> Rc<T> {
+        self
+    }
+
+    fn cast_arc_to(self: Arc<Self>) -> Arc<T> {
+        self
+    }
+}
+
+pub trait Foo {
+    fn foo(&self) {
+        println!("Foo({})", core::any::type_name::<Self>());
+    }
+}
+
+pub trait Bar: CastTo<dyn Foo> + CastTo<dyn core::fmt::Debug> + CastTo<[i32]> {
+    fn bar(&self) {
+        println!("Bar({})", core::any::type_name::<Self>());
+    }
+}
+
+impl Foo for [i32; 10] {}
+impl Bar for [i32; 10] {}
+
+fn main() {
+    let x = [0; 10];
+    let x: Box<dyn Bar> = Box::new(x);
+    let x = (*x).cast::<[i32]>();
+    //~^ ERROR: the trait bound `dyn Bar: CastTo<[i32]>` is not satisfied
+}

--- a/src/test/ui/unsized/issue-71659.stderr
+++ b/src/test/ui/unsized/issue-71659.stderr
@@ -1,0 +1,9 @@
+error[E0277]: the trait bound `dyn Bar: CastTo<[i32]>` is not satisfied
+  --> $DIR/issue-71659.rs:92:18
+   |
+LL |     let x = (*x).cast::<[i32]>();
+   |                  ^^^^ the trait `CastTo<[i32]>` is not implemented for `dyn Bar`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0277`.

--- a/src/test/ui/unsized/issue-71659.stderr
+++ b/src/test/ui/unsized/issue-71659.stderr
@@ -1,8 +1,8 @@
-error[E0277]: the trait bound `dyn Bar: CastTo<[i32]>` is not satisfied
-  --> $DIR/issue-71659.rs:92:18
+error[E0277]: the trait bound `dyn Foo: CastTo<[i32]>` is not satisfied
+  --> $DIR/issue-71659.rs:30:15
    |
-LL |     let x = (*x).cast::<[i32]>();
-   |                  ^^^^ the trait `CastTo<[i32]>` is not implemented for `dyn Bar`
+LL |     let x = x.cast::<[i32]>();
+   |               ^^^^ the trait `CastTo<[i32]>` is not implemented for `dyn Foo`
 
 error: aborting due to previous error
 

--- a/src/test/ui/unsized/issue-75707.rs
+++ b/src/test/ui/unsized/issue-75707.rs
@@ -1,0 +1,17 @@
+pub trait Callback {
+    fn cb();
+}
+
+pub trait Processing {
+    type Call: Callback;
+}
+
+fn f<P: Processing + ?Sized>() {
+    P::Call::cb();
+}
+
+fn main() {
+    struct MyCall;
+    f::<dyn Processing<Call = MyCall>>();
+    //~^ ERROR: the trait bound `MyCall: Callback` is not satisfied
+}

--- a/src/test/ui/unsized/issue-75707.stderr
+++ b/src/test/ui/unsized/issue-75707.stderr
@@ -1,0 +1,12 @@
+error[E0277]: the trait bound `MyCall: Callback` is not satisfied
+  --> $DIR/issue-75707.rs:15:5
+   |
+LL | fn f<P: Processing + ?Sized>() {
+   |         ---------- required by this bound in `f`
+...
+LL |     f::<dyn Processing<Call = MyCall>>();
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Callback` is not implemented for `MyCall`
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0277`.

